---
layout: default
title: {{ site.name }}
---

<p><strong>alntools</strong> processes next-generation sequencing read alignments into a sparse compressed incidence matrix (aka Equivalence Classes) and stores it in a pre-defined <a href="#output">binary format</a> for efficient downstream analyses and storage. It enables us to compare, contrast, or combine the results of different alignment strategies. </p>

<ul><li>Free software: GPLv3 license</li></ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p><strong>Note:</strong> Although alntools is available at PyPI for <code>pip install</code> or <code>easy_install</code>, we highly recommend using <a href="https://www.continuum.io/downloads">Anaconda distribution</a> of python to install all the dependencies without issues. We are going to post alntools on <a href="https://anaconda.org/kbchoi">Anaconda Cloud</a> soon, so add the following channels if you have not already:</p>

<pre>
  <code>$ conda config --add channels r
  $ conda config --add channels bioconda</code>
</pre>

<p>To avoid conflicts among dependencies, we also highly recommend using conda virtual environment and install alntools in it:</p>

<pre>
  <code>$ git clone https://github.com/churchill-lab/alntools
  $ cd alntools
  $ conda create -n alntools python=2
  $ source activate alntools
  (alntools) $ conda install -c kbchoi emase
  (alntools) $ conda install future
  (alntools) $ python setup.py install</code>
</pre>

<p>That's all! Now alntools is ready. Once you are done using alntools, you can go out from its virtual environment anytime:</p>

<pre><code>(alntools) $ source deactivate
</code></pre>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<pre><code>$ alntools -h
Usage: alntools [OPTIONS] COMMAND [ARGS]...

Commands:
  split      split a BAM file into chunks
  range      Report the range of alignment to each target
  bam2ec     Convert a BAM file to EC
  bam2emase  Convert a BAM file to APM
  ec2emase   Convert an EC file to EMASE
  emase2ec   Convert an EMASE file to EC

Options:
  --version   Show the version and exit.
  -h, --help  Show this message and exit.</code>
</pre>

<p>&#8226;<code><strong>split</strong></code> divides a bam file into smaller bam files (with the same header) to help parallelism in any downstream analyses. An input bam file is split such that all the alignment of a same read appear only in a single chunk. No sorting required.</p>

<pre><code>$ alntools split -h
Usage: alntools split &lt;options&gt; bam_file N

  Splits a BAM file (bam_file) into N smaller ones

Options:
  -c, --chunks INTEGER       Number of chunks to process
  -d, --directory DIRECTORY  Output directory for storing bam file chunks
  -v, --verbose              Run in verbose mode
  -h, --help                 Show this message and exit.</code>
</pre>

<p>&#8226;<code><strong>range</strong></code> reports <em>the effective lengths</em> of alignment targets from the input data by finding the left and rightmost base of all reads that align to each target.</p>

<pre><code>$ alntools range -h
Usage: alntools range &lt;options&gt; [INPUT]... range_file

  Create range file for specified BAM files

Options:
  -d, --directory DIRECTORY  temp directory
  -t, --targets TARGET FILE  Target ID file
  -v, --verbose              Run in verbose mode
  -h, --help                 Show this message and exit.</code>
</pre>

<p>&#8226;<code><strong>bam2ec</strong></code> processes a bam file, creates equivalence classes, and exports them in a <a href="#output">binary format</a>. It runs in parallel by dividing an input bam file into chunks. If <code>--range</code> option is set, it reports effective target lengths according to alignment data.</p>
<pre>
<code>$ alntools bam2ec -h
Usage: alntools bam2ec &lt;options&gt; bam_file ec_file

  Convert a BAM file (bam_file) to an EC file (ec_file).

Options:
  -c, --chunks INTEGER       Number of chunks to process
  -d, --directory DIRECTORY  Temp directory to store bam file chunks
  --range PATH               Report alignment range
  -t, --targets TARGET FILE  Target ID file
  -v, --verbose              Run in verbose mode
  -h, --help                 Show this message and exit.</code>
</pre>

<p>&#8226;<code><strong>bam2emase</strong></code>processes a bam file, creates equivalence classes, and exports them in an <a href="http://emase.readthedocs.io/en/latest/readme.html">emase</a> HDF5 format so we can use EMASE with equivalence classes. It also runs in parallel by dividing an input bam file into chunks.</p>
<pre><code>$ alntools bam2emase -h
Usage: alntools bam2emase &lt;options&gt; bam_file emase_file

  Convert a BAM file (bam_file) to an EMASE file (emase_file).

Options:
  -c, --chunks INTEGER       Number of chunks to process
  -d, --directory DIRECTORY  Temp directory to store bam file chunks
  -t, --targets TARGET FILE  Target ID file
  -v, --verbose              Run in verbose mode
  -h, --help                 Show this message and exit.</code>
</pre>


<p><a id="ec2emase" href="#ec2emase"></a>&#8226;<code><strong>ec2emase</strong></code> and <code><strong>emase2ec</strong></code> support conversion between binary and emase formats of equivalence class file.</p>
<pre><code>$ alntools ec2emase -h
Usage: alntools ec2emase &lt;options&gt; ec_file emase_file

  Convert an binary EC file (ec_file) to an EMASE file (emase_file).

Options:
  -v, --verbose  Run in verbose mode
  -h, --help     Show this message and exit.</code>
</pre>
<pre><code>$ alntools emase2ec -h
Usage: alntools emase2ec &lt;options&gt; emase_file ec_file

  Convert an EMASE file (emase_file) to an binary EC file (ec_file).

Options:
  -v, --verbose  Run in verbose mode
  -h, --help     Show this message and exit.</code>
</pre>


<h2>
<a id="output" class="anchor" href="#output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output format</h2>

<h4>Binary (EC) format</h4>
<p><img src="emase_binary_format.jpg" alt="" title="Binary EC format"></p>

<h4>EMASE format</h4>
<p><img src="emase_hdf5_format.jpg" alt="" title="EMASE HDF5 format"></p>

<br>
<div id="home">
  <h2>Announcements</h2>
  <ul class="posts">
    {% for post in site.posts %}
      <li><span>{{ post.date | date_to_string }}</span> &raquo; <a href="{{ site.baseurl }}{{ post.url }}">{{ post.title }}</a></li>
    {% endfor %}
  </ul>
</div>
